contracts/cw20-streams/src/contract.rs

```diff
@@ -0,0 +1,446 @@
+use crate::error::ContractError;
+use crate::msg::{
+    ConfigResponse, ExecuteMsg, InstantiateMsg, QueryMsg, ReceiveMsg, StreamResponse,
+};
+use crate::state::{save_stream, Config, Stream, CONFIG, STREAMS, STREAM_SEQ};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    from_binary, to_binary, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw20::{Cw20Contract, Cw20ExecuteMsg, Cw20ReceiveMsg};
+
+const CONTRACT_NAME: &str = "crates.io:cw20-streams";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = msg
+        .owner
+        .and_then(|s| deps.api.addr_validate(s.as_str()).ok())
+        .unwrap_or(info.sender);
+    let config = Config {
+        owner: owner.clone(),
+        cw20_addr: deps.api.addr_validate(msg.cw20_addr.as_str())?,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    STREAM_SEQ.save(deps.storage, &0u64)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "instantiate")
+        .add_attribute("owner", owner)
+        .add_attribute("cw20_addr", msg.cw20_addr))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => execute_receive(env, deps, info, msg),
+        ExecuteMsg::Withdraw { id } => try_withdraw(env, deps, info, id),
+    }
+}
+
+pub fn try_create_stream(
+    env: Env,
+    deps: DepsMut,
+    owner: String,
+    recipient: String,
+    amount: Uint128,
+    start_time: u64,
+    end_time: u64,
+) -> Result<Response, ContractError> {
+    let validated_owner = deps.api.addr_validate(owner.as_str())?;
+    if validated_owner != owner {
+        return Err(ContractError::InvalidOwner {});
+    }
+
+    let validated_recipient = deps.api.addr_validate(recipient.as_str())?;
+    if validated_recipient != recipient {
+        return Err(ContractError::InvalidRecipient {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    if config.owner == recipient {
+        return Err(ContractError::InvalidRecipient {});
+    }
+
+    if start_time > end_time {
+        return Err(ContractError::InvalidStartTime {});
+    }
+
+    let block_time = env.block.time.seconds();
+    if start_time < block_time {
+        return Err(ContractError::InvalidStartTime {});
+    }
+
+    let duration: Uint128 = end_time.checked_sub(start_time).unwrap().into();
```

Comment:
Removed the unwrap code. Was a bit unsure about the best way to implement checked arithmetic in a clean way, so unwrap seemed simplest. But much more aware now.

Also checked_sub actually isn't needed here since we throw an error earlier if start_time > end_time. So just used regular subtraction here.